pkg/controllers/proxy/network_services_controller.go

```diff
@@ -1629,17 +1629,20 @@ func (nsc *NetworkServicesController) setupMangleTableRule(ip string, protocol s
 	}
 
 	// setup iptables rule TCPMSS for DSR mode to fix mtu problem
-	mtuArgs := []string{"-d", ip, "-m", tcpProtocol, "-p", tcpProtocol, "--tcp-flags", "SYN,RST", "SYN", "-j", "TCPMSS",
-		"--set-mss", strconv.Itoa(tcpMSS)}
-	err = iptablesCmdHandler.AppendUnique("mangle", "PREROUTING", mtuArgs...)
-	if err != nil {
-		return fmt.Errorf("failed to run iptables command to set up TCPMSS due to %v", err)
-	}
-	mtuArgs[0] = "-s"
-	err = iptablesCmdHandler.AppendUnique("mangle", "POSTROUTING", mtuArgs...)
-	if err != nil {
-		return fmt.Errorf("failed to run iptables command to set up TCPMSS due to %v", err)
+	if protocol == tcpProtocol {
+		mtuArgs := []string{"-d", ip, "-m", tcpProtocol, "-p", tcpProtocol, "--dport", port,
+			"--tcp-flags", "SYN,RST", "SYN", "-j", "TCPMSS", "--set-mss", strconv.Itoa(tcpMSS)}
+		err = iptablesCmdHandler.AppendUnique("mangle", "PREROUTING", mtuArgs...)
+		if err != nil {
+			return fmt.Errorf("failed to run iptables command to set up TCPMSS due to %v", err)
+		}
+		mtuArgs[0] = "-s"
```

Comment:
Wouldn't you need to change `--dport` to `--sport` here as well as changing the `-d` to `-s`?

Reply:
You are right, I've missed that. Howewer, according to [TCP](https://datatracker.ietf.org/doc/html/rfc9293#name-maximum-segment-size-option) it is enough to reduce MSS in the first SYN packet. On the other hand, I miss a similar rule in OUTPUT chain in table mangle, which I will add that soon after some testing.

Reply:
Unfortunately, reducing MSS in intiail SYN packets is not enough, at least linux implementation responds with a higher MSS value. According to kube-router's design, this way the sender side dont get any clue of the reduced MSS. In theory, it is an assymetric routing, where only the connecting path is affected, thus it would be enough to just reduce MSS on SYN+ACK reply packages. I need more research and experiment on this.